core/src/main/scala/kafka/log/Log.scala

```diff
@@ -181,25 +191,38 @@ class Log(val dir: File,
         // if its a log file, load the corresponding log segment
         val start = filename.substring(0, filename.length - LogFileSuffix.length).toLong
         val indexFile = Log.indexFilename(dir, start)
+        val timeIndexFile = Log.timeIndexFilename(dir, start)
         val segment = new LogSegment(dir = dir,
                                      startOffset = start,
                                      indexIntervalBytes = config.indexInterval,
                                      maxIndexSize = config.maxIndexSize,
                                      rollJitterMs = config.randomSegmentJitter,
                                      time = time,
                                      fileAlreadyExists = true)
+        if (indexFile.exists() && timeIndexFile.exists()) {
 
-        if(indexFile.exists()) {
           try {
-              segment.index.sanityCheck()
+            segment.index.sanityCheck()
           } catch {
             case e: java.lang.IllegalArgumentException =>
-              warn("Found a corrupted index file, %s, deleting and rebuilding index...".format(indexFile.getAbsolutePath))
+              warn(s"Found a corrupted index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                s"${indexFile.getAbsolutePath} and rebuilding index...")
               indexFile.delete()
+              timeIndexFile.delete()
               segment.recover(config.maxMessageSize)
           }
-        }
-        else {
+
+          try {
+            segment.timeIndex.sanityCheck()
+          } catch {
+            case e: java.lang.IllegalArgumentException =>
+              warn(s"Found a corrupted time index file, deleting ${timeIndexFile.getAbsolutePath}, " +
+                  s"${indexFile.getAbsolutePath} and rebuilding index...")
+              indexFile.delete()
+              timeIndexFile.delete()
+              segment.recover(config.maxMessageSize)
+          }
+        } else {
```

Comment:
The current code actually creates the time index file in line 195 when it instantiate the LogSegment. So the else branch seems never be called. I asked that question earlier and you suggested that we can check the index existence before we instantiate the LogSegment. But the problem of that is we are not able to distinguish between 1) upgrade case and 2) the time index is missing. We want to create an empty index in case 1) and want to rebuild the time index in case 2).

I thought case 2) only occurs when the time index is manually deleted, which is rare. Therefore I left the code as it is, i.e. always create an empty time index when it is missing. This is essentially assuming that when a time index is missing it is the upgrade case.